lib/rubygems/spec_fetcher.rb

```diff
@@ -174,20 +174,35 @@ def suggest_gems_from_name(gem_name, type = :latest, num_results = 5)
     max             = gem_name.size / 2
     names           = available_specs(type).first.values.flatten(1)
 
+    min_length = gem_name.length - max
+    max_length = gem_name.length + max
+
     matches = names.map do |n|
+      len = n.name.length
+      # If the length is min_length or shorter, we've done `max` deletions.
+      # If the length is max_length or longer, we've done `max` insertions.
+      # These would both be rejected later, so we skip early for performance.
+      next if len < min_length || len > max_length
+
+      # If the gem doesn't support the current platform, bail early.
       next unless n.match_platform?
-      [n.name, 0] if n.name.downcase.tr("_-", "").include?(gem_name)
-    end.compact
 
-    if matches.length < num_results
-      matches += names.map do |n|
-        next unless n.match_platform?
-        distance = levenshtein_distance gem_name, n.name.downcase.tr("_-", "")
-        next if distance >= max
-        return [n.name] if distance == 0
-        [n.name, distance]
-      end.compact
-    end
+      # The candidate name, normalized the same as gem_name.
+      normalized_name = n.name.downcase.tr("_-", "")
```

Comment:
The old behavior was to collect all gems where removing dash and underscore matches the same gem. The new implementation just returns the first match.

```sh
$ gem install active-support-addon
did you mean: activesupport_addon, active_support-addon
```

You'll need to `next [n.name, 0]` to prevent hiding other gems with the same name when hyphens/dashes are removed.

When the gem name input is 3 characters or less, we don't allow any distance for matches. I'm inclined to suggest that even a hyphen removed match for a 3 character gem name is not very useful. Is there a gem we should suggest named `ab_c` when someone enters `abc`? If so, we need to short circuit short gem input names after this. If not, then we could skip the entire loop and return nothing for short gem names.

Reply:
I didn't _add_ the return, I just modified it to not require running `levenshtein_distance` when it wasn't needed.

This is the current code on `master`:

https://github.com/rubygems/rubygems/blob/74a8893640bce601a7feb5c164374166eb989e7a/lib/rubygems/spec_fetcher.rb#L177-L183

It would be a behavior change.

Reply:
You're right. I misread.